Move Server to Deployment instead of DaemonSet

daviddyball opened the pull request:
See discussion in [this issue](https://github.com/uswitch/kiam/issues/245)

Considerations:
* I've set `replicas` to 3, but most will want to adjust this
* I've assumed that antiAffinity is based on pod co-location and not some other criteria
* The `nodeSelector` still ensures it only runs on `master`s, while `spec.affinity` prevents co-location on the same node
* Would it be worth adding some YAML for Horizontal Pod Autoscaling at this stage?

EDIT: Also relates to [issues rolling master nodes causing issues](https://github.com/uswitch/kiam/issues/217)

daviddyball commented:
@rhysemmas @Joseph-Irving : I saw that you had recent approval activity on some closed PRs. Are you active maintainers? Are you able to review this PR?

stefansedich commented:
Would love to see this one merged, however I wonder @daviddyball if it should provide an option to either run as a DaemonSet or a Deployment depending on what a user prefers.

daviddyball commented:
Very true @stefansedich. It's currently not a helm template, so adding some sort of "switching" logic is not really possible. Perhaps just offering two different YAML files is enough? My only concern would be people blindly doing `kubectl apply -f deploy/*.yaml` :stuck_out_tongue:

stefansedich commented:
Ah sorry @daviddyball it is early and I failed to realize this was not the helm chart!

denniswebb commented:
If this has gotten stale, I'm open to doing the helm chart work because as it stands, I'm about to have to start hosting the chart myself to do the daemonset vs deployment work. Please ping me if I'm open to do the helm chart work myself on a separate PR.

daviddyball commented:
This isn't exactly stale, it's just not gotten reviewed by any maintainers :confused: We've been running `kiam-server` as a `Deployment` since September and it's solved most of our issues with masters being rolled.

jaygorrell commented:
I'm curious if the maintainers have any opposition to this approach. The chart PR linked above (https://github.com/uswitch/kiam/pull/332) adds an _option_ for moving to Deployment without changing the default... so hopefully that can get accepted more easily.

If the default is changing, the chart should change to match but also there are things like the README that should be updated as part of this PR. We just need input from a maintainer though.

daviddyball commented:
This one doesn't force a default, it just offers a different YAML file if you _choose_ to run it as a `Deployment` :man_shrugging:

jaygorrell commented:
It's not a different YAML file though - it's changing the `DaemonSet` file to be `Deployment`. 

I have no authority here, just saying it may be confusing to have the README talk about using a `DaemonSet` while the example in `deploy/server.yaml` is a `Deployment`. I like the change, but updating the README would make sense.

daviddyball commented:
I had it in my head that I'd already refactored this to make it optional which deployment method you choose... turns out I hadn't :roll_eyes: 

Moved to 2x YAML file options under `deploy/`, `server.daemonset.yaml` and `server.deployment.yaml` and updated `README.md` to include instructions and a note about only deploying _one_ of the two options available.

jaygorrell commented:
Looks good to me. I'd approve if I could. :)

Joseph-Irving commented on README.md:
````diff
@@ -63,6 +63,18 @@ When your process starts an AWS SDK library will normally use a chain of credent
+```
+kubectl apply -f deploy/server.daemonset.yaml
+kubectl apply -f deploy/service.yaml
+```
````
I would probably avoid directly telling people to kubectl apply these manifests, they were intended as examples. They're not for production use, like they use latest as the tag and potentially need you to change some things such as where you mount certs in from. So I think I'd just call out that there are two ways to deploy and you can find them in the deploy folder if you want examples

jaygorrell replied:
Not that I disagree, but this does contextually fit with what was already here.

daviddyball replied:
Removed

dbyron0 commented on deploy/server.deployment.yaml:
```diff
@@ -0,0 +1,99 @@
+    role: server
+spec:
+  replicas: 3
+  selector:
```
I think this is missing:
```
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
```
as the helm chart has [here](https://github.com/uswitch/kiam/blob/master/helm/kiam/templates/server-deployment.yaml#L14)

See https://github.com/uswitch/kiam/pull/332/files#r351312455 for background.
